base, configs, mem, python, stdlib: Refactor addr range

powerjg opened the pull request:
See #2855 for more details. Detailed commit messages in each commit.

**Important note**: Gemini wrote most of this code. I supervised, tested, and carefully reviewed.

powerjg committed:
base: Add policy to AddrRange

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: Add modulo-based AddrRange

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Expose Modulo AddrRange to python

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: Add sparse AddrRange policy

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: addrrange improvements

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: Updates for addr range

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Add ModuloAddrRange class

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: Add print for ModuloAddrRangePolicy

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python,mem: Add support for SparseAddrRanges to memory

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
stdlib: Use SparseAddrRange for x86 board

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base,mem: Fix createBackingStore

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
python: Fix printing sparse ranges

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
mem: Fix abstract memory

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
stdlib: Add e820 entry to x86board

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
Output from `/proc/meminfo` and early printk:

```
[    0.000000] BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x000000000009fc00-0x00000000000fffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x00000000bfffffff] usable
[    0.000000] BIOS-e820: [mem 0x00000000ffff0000-0x00000000ffffffff] reserved
[    0.000000] BIOS-e820: [mem 0x0000000100000000-0x000000083fffffff] usable

...

In after_boot.sh...
Interactive mode: false
Starting gem5 init... trying to read run script file via readfile.
Running script from gem5-bridge stored in /tmp/script
MemTotal:       32868444 kB
MemFree:        32626468 kB
MemAvailable:   32429096 kB
Buffers:           11316 kB
Cached:           114844 kB
SwapCached:            0 kB
Active:           141272 kB
Inactive:          24028 kB
Active(anon):      39660 kB
Inactive(anon):        0 kB
Active(file):     101612 kB
Inactive(file):    24028 kB
Unevictable:           0 kB
Mlocked:               0 kB
SwapTotal:             0 kB
SwapFree:              0 kB
Zswap:                 0 kB
Zswapped:              0 kB
Dirty:              1712 kB
Writeback:           148 kB
AnonPages:         39188 kB
Mapped:            53504 kB
Shmem:               368 kB
KReclaimable:       9876 kB
Slab:              32444 kB
SReclaimable:       9876 kB
SUnreclaim:        22568 kB
KernelStack:        1760 kB
PageTables:         1780 kB
SecPageTables:         0 kB
NFS_Unstable:          0 kB
Bounce:                0 kB
WritebackTmp:          0 kB
CommitLimit:    16434220 kB
Committed_AS:     176872 kB
VmallocTotal:   34359738367 kB
VmallocUsed:        8736 kB
VmallocChunk:          0 kB
Percpu:              864 kB
HardwareCorrupted:     0 kB
AnonHugePages:         0 kB
ShmemHugePages:        0 kB
ShmemPmdMapped:        0 kB
FileHugePages:         0 kB
FilePmdMapped:         0 kB
Unaccepted:            0 kB
HugePages_Total:       0
HugePages_Free:        0
HugePages_Rsvd:        0
HugePages_Surp:        0
Hugepagesize:       2048 kB
Hugetlb:               0 kB
DirectMap4k:       40960 kB
DirectMap2M:    33513472 kB
```

Tests to see if the address mapping still works. Here are some of the stats. Each channel has about the same page hit rate, and nearly the same read/write requests. Also, it looks like the accesses are mostly distributed evenly across the banks. This test was run with kvm and using the timing CPU after boot. I compared to the old version and things look *similar*, but not exactly the same.

```
board.memory.mem_ctrl0.dram.readRowHitRate        60.72                       # Row buffer hit rate for reads (Ratio)
board.memory.mem_ctrl1.dram.readRowHitRate        61.34                       # Row buffer hit rate for reads (Ratio)

board.memory.mem_ctrl0.readReqs                274240                       # Number of read requests accepted (Count)
board.memory.mem_ctrl1.readReqs                263593                       # Number of read requests accepted (Count)
board.memory.mem_ctrl0.writeReqs               102066                       # Number of write requests accepted (Count)
board.memory.mem_ctrl1.writeReqs                98626                       # Number of write requests accepted (Count)

board.memory.mem_ctrl0.dram.perBankRdBursts::0         7183                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::1         7022                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::2         9854                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::3         7266                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::4         9030                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::5         6440                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::6         9289                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::7         7044                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::8        10356                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::9         8481                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::10         7521                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::11         7805                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::12         9397                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::13         9463                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::14         6371                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::15         9020                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::16         9224                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::17         7284                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::18         8310                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::19         8686                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::20         9741                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::21         9078                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::22         7825                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::23        10640                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::24         8991                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::25         8475                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::26         8044                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::27         9164                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::28         9005                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::29         8571                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::30         8445                       # Per bank write bursts (Count)
board.memory.mem_ctrl0.dram.perBankRdBursts::31         9304                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::0         6679                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::1         7050                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::2         9494                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::3         7271                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::4         8881                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::5         6021                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::6         8711                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::7         6788                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::8        10239                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::9         8221                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::10         7152                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::11         7612                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::12         8562                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::13         8361                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::14         6155                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::15         8700                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::16         8889                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::17         6783                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::18         7659                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::19         8377                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::20         9186                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::21         8646                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::22         7709                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::23        10212                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::24         8902                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::25         8328                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::26         7776                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::27         9053                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::28         8856                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::29         8055                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::30         8337                       # Per bank write bursts (Count)
board.memory.mem_ctrl1.dram.perBankRdBursts::31         9161                       # Per bank write bursts (Count)
```

powerjg committed:
base: Fix case of modulo policy not working with 1 channel

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base,python: Add docs to AddrRange constructors

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
mem: Optimize isMemAddr

This function is on the critical path, especially in Ruby.
It takes about 2% of the total cycles during linux boot with
the timing CPU. To optimize, we make a very simple datastructure
in the PhysicalMemory object and query that instead of
using the AddrRangeMap.

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
mem,base: Update PhysicalMemory to correctly handle sparse ranges

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base: Move AddrRange sparse code to AddrRange

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base,mem: Fixing bugs

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
base,mem,stdlib: Cleanup PR

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
I've tested booting x86 with dual channel + sparse ranges on both KVM and timing CPUs. I've also tested booting Arm with dual channel (not sparse). Everything seems to be working. I know it's a big change. I still would like to cleanup the commits. Please provide any feedback before I do, though!

powerjg committed:
configs: Remove x86 board 3GiB memory limitation

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
@abmerop FYI, I did not update the ViperBoard.

powerjg committed:
base: Add missing header to addr_range_policy

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg committed:
mem: Fix unintialized variable

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
@abmerop or @mattsinc any idea why this would cause the SE GPU script to fail? I would love a pointer... It's quite difficult for me to run those tests locally. I don't have access to docker easily.

The failure is

```
gem5.opt: src/mem/ruby/common/NetDest.cc:55: void gem5::ruby::NetDest::add(gem5::ruby::MachineID): Assertion `bitIndex(newElement.num) < m_bits[vecIndex(newElement)].getSize()' failed.
```

Given this is SE mode, I don't know what would have changed in this PR. All ranges are simple uninterleaved ranges, from what I can tell.

abmerop commented:
> @abmerop or @mattsinc any idea why this would cause the SE GPU script to fail? I would love a pointer... It's quite difficult for me to run those tests locally. I don't have access to docker easily.
> 
> The failure is
> 
> ```
> gem5.opt: src/mem/ruby/common/NetDest.cc:55: void gem5::ruby::NetDest::add(gem5::ruby::MachineID): Assertion `bitIndex(newElement.num) < m_bits[vecIndex(newElement)].getSize()' failed.
> ```
> 
> Given this is SE mode, I don't know what would have changed in this PR. All ranges are simple uninterleaved ranges, from what I can tell.

Not sure off the top of my head.  If Matt has no idea either I can take a look Sunday at the latest.  Have you tested if GPUFS has the same issue?  I kind of expect it would given they use the same Ruby protocol

mattsinc commented:
> > @abmerop or @mattsinc any idea why this would cause the SE GPU script to fail? I would love a pointer... It's quite difficult for me to run those tests locally. I don't have access to docker easily.
> > The failure is
> > ```
> > gem5.opt: src/mem/ruby/common/NetDest.cc:55: void gem5::ruby::NetDest::add(gem5::ruby::MachineID): Assertion `bitIndex(newElement.num) < m_bits[vecIndex(newElement)].getSize()' failed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Given this is SE mode, I don't know what would have changed in this PR. All ranges are simple uninterleaved ranges, from what I can tell.
> 
> Not sure off the top of my head. If Matt has no idea either I can take a look Sunday at the latest. Have you tested if GPUFS has the same issue? I kind of expect it would given they use the same Ruby protocol

@v-ramadas : I think we've seen this before when we were changing the HBM addressing scheme for @TomXia 's patches?  But I can't recall what we changed.

powerjg committed:
mem: Fix incorrect assert

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
> Have you tested if GPUFS has the same issue?

No, but I'd like to. Do we have prebuilt disks/kernels/workloads? I can't seem to find them on resources.gem5.org.

powerjg committed:
stdlib: Fix HBMStack to work with updated memory

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
I fixed the issue with the HBMStack. I hope that fixes the GPU issue, but I doubt it.

mattsinc commented:
> > > @abmerop or @mattsinc any idea why this would cause the SE GPU script to fail? I would love a pointer... It's quite difficult for me to run those tests locally. I don't have access to docker easily.
> > > The failure is
> > > ```
> > > gem5.opt: src/mem/ruby/common/NetDest.cc:55: void gem5::ruby::NetDest::add(gem5::ruby::MachineID): Assertion `bitIndex(newElement.num) < m_bits[vecIndex(newElement)].getSize()' failed.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Given this is SE mode, I don't know what would have changed in this PR. All ranges are simple uninterleaved ranges, from what I can tell.
> > 
> > 
> > Not sure off the top of my head. If Matt has no idea either I can take a look Sunday at the latest. Have you tested if GPUFS has the same issue? I kind of expect it would given they use the same Ruby protocol
> 
> @v-ramadas : I think we've seen this before when we were changing the HBM addressing scheme for @TomXia 's patches? But I can't recall what we changed.

@v-ramadas , @TomXia reminder to please chime in here

abmerop commented:
> I fixed the issue with the HBMStack. I hope that fixes the GPU issue, but I doubt it.

Yea, I am now seeing an issue on the first PCI access. I haven't done too much digging yet, but how does this PR handle address ranges like PCI BARs that are unknown at gem5 config time? These would be set when the simulated OS sets them at runtime. I recall a mention somewhere about moving away from Ruby[Port]'s assumption that anything outside the main memory range is assumed to be I/O. Is that part of this PR or am I confusing with a different PR?

TomXia commented:
> > > > @abmerop or @mattsinc any idea why this would cause the SE GPU script to fail? I would love a pointer... It's quite difficult for me to run those tests locally. I don't have access to docker easily.
> > > > The failure is
> > > > ```
> > > > gem5.opt: src/mem/ruby/common/NetDest.cc:55: void gem5::ruby::NetDest::add(gem5::ruby::MachineID): Assertion `bitIndex(newElement.num) < m_bits[vecIndex(newElement)].getSize()' failed.
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Given this is SE mode, I don't know what would have changed in this PR. All ranges are simple uninterleaved ranges, from what I can tell.
> > > 
> > > 
> > > Not sure off the top of my head. If Matt has no idea either I can take a look Sunday at the latest. Have you tested if GPUFS has the same issue? I kind of expect it would given they use the same Ruby protocol
> > 
> > 
> > @v-ramadas : I think we've seen this before when we were changing the HBM addressing scheme for @TomXia 's patches? But I can't recall what we changed.
> 
> @v-ramadas , @TomXia reminder to please chime in here

Apologies for the delay. I do recall encountering errors previously due to an incorrectly configured interleaving scheme. However, this specific assertion appears to be new and different from the ones we’ve seen before.

powerjg commented:
> Yea, I am now seeing an issue on the first PCI access.

I *thought* I had tested GPU FS and things seemed to work. I'll double check this.

> I haven't done too much digging yet, but how does this PR handle address ranges like PCI BARs that are unknown at gem5 config time? These would be set when the simulated OS sets them at runtime.

This should be completely unchanged. I don't think the simulator creates any address ranges on the fly for this, but I could be mistaken. If I'm wrong, do you know where the code is for this?

> I recall a mention somewhere about moving away from Ruby[Port]'s assumption that anything outside the main memory range is assumed to be I/O. Is that part of this PR or am I confusing with a different PR?

I optimized this path in the Ruby port, but the logic is unchanged.

abmerop commented:
> > Yea, I am now seeing an issue on the first PCI access.
> 
> I _thought_ I had tested GPU FS and things seemed to work. I'll double check this.

I currently run ~70 tests and they all fail the same way. It seems to not be finding a peer port (based on invalid this pointers):

```
#0  0x000055555943f286 in gem5::RequestPort::sendAtomic (this=0x348, pkt=0x55555f387c30) at src/mem/port.hh:556
#1  0x000055555b8ad4bb in gem5::ruby::AbstractController::recvAtomic (this=0x0, pkt=0x55555f387c30) at src/mem/ruby/slicc_interface/AbstractController.cc:422
#2  0x000055555b511db5 in gem5::ruby::RubyPort::MemResponsePort::recvAtomic (this=0x55555ea4ad80, pkt=0x55555f387c30) at src/mem/ruby/system/RubyPort.cc:393
#3  0x000055555b8bb040 in gem5::AtomicRequestProtocol::send (this=0x55555d969ad8, peer=0x55555ea4adc0, pkt=0x55555f387c30) at src/mem/protocol/atomic.cc:54
#4  0x000055555943f2b1 in gem5::RequestPort::sendAtomic (this=0x55555d969a98, pkt=0x55555f387c30) at src/mem/port.hh:556
#5  0x0000555559d70128 in gem5::BaseKvmCPU::KVMCpuPort::submitIO (this=0x55555d969a98, pkt=0x55555f387c30) at src/cpu/kvm/base.cc:194
...
```

powerjg commented:
I am able to reproduce the error with GPU FS. So, I'll work on debugging it.

However, it looks like something is misconfigured on the disk image. I'm seeing this output.

```
Can't open /dev/gem5_bridge: No such file or directory
--> Make sure the gem5_bridge device driver has been properly inserted into the kernel. Otherwise, sudo access required to perform address-mode ops when linking against m5 library.
3+0 records in
3+0 records out
3072 bytes (3.1 kB, 3.0 KiB) copied, 0.00599909 s, 512 kB/s
[   14.075923] amdgpu 0000:00:08.0: amdgpu: Unhandled NPS info table 0.0
No GPU available.
tensor([[0.5402, 0.0949, 0.8380],
        [0.6795, 0.6509, 0.5943],
        [0.0825, 0.1875, 0.7475],
        [0.2452, 0.8151, 0.7634],
        [0.5808, 0.5483, 0.3453]])
Can't open /dev/gem5_bridge: No such file or directory
--> Make sure the gem5_bridge device driver has been properly inserted into the kernel. Otherwise, sudo access required to perform address-mode ops when linking against m5 library.
```

powerjg committed:
mem: Fix boundary condition in isMemAddr

Address ranges are exclusive on the max address. Update
isMemAddr to do the same check

Signed-off-by: Jason Lowe-Power <jlowepower@google.com>

powerjg commented:
Thanks for the help! I found the issue in GPUFS. I hope that fixes the SE GPU issue too. I now get the same behavior on the develop branch and on this branch.

powerjg commented:
🎉  everything is passing. 

@giactra, @abmerop, @BobbyRBruce let me know if there's anything I can do to make reviewing easier or if you have any questions.

BobbyRBruce commented on src/base/addr_range.hh:
```diff
@@ -89,69 +92,65 @@ class AddrRange
 
   public:
-
     /**
+     * Default constructor, creates an invalid address range.
+     *
      * @ingroup api_addr_range
      */
-    AddrRange()
-        : _start(1), _end(0), intlvMatch(0)
-    {}
```
Is there a reason we'd ever want  to use this constructor? Shouldn't the default constructor just be disabled?

powerjg replied:
It's used in a few places in the codebase. I believe it's also used in the AddrRangeMap class (but I could be mistaken there).

BobbyRBruce replied:
I took a look into this. It’s only used in two places in the codebase: ["src/mem/backdoor.hh" at line 96](https://github.com/gem5/gem5/blob/7a2b0e413d06c5ce7097104abef3b1d9eaabca91/src/mem/backdoor.hh#L96) and ["src/dev/amdgpu/amdgpu_vm.cc" at line 54](https://github.com/gem5/gem5/blob/7a2b0e413d06c5ce7097104abef3b1d9eaabca91/src/dev/amdgpu/amdgpu_vm.cc). In both cases, it’s used as a "Not set" option. `contains()` and `valid` always return `false`, which is what you’d expect from such an option. It would probably be better to have a special constructor for the "Not set" case to make it safer and force the user to state explicitly it's what they want. If we were to refactor this I'd suggest a factory like `AddrRange::not_set()` with this default constructor disabled. 

However, the constructors’ usage in "backdoor.hh" dates back 7 years, and it's technically an API function we'd need to deprecate before removing, so I won't push for it in this PR.

BobbyRBruce commented on src/base/addr_range.hh:
```diff
@@ -182,13 +181,124 @@ class AddrRange
+     *
+     * @ingroup api_addr_range
+     */
+    AddrRange(const std::vector<std::pair<Addr, Addr>> &ranges)
```
It's fine this way, so not a hard requirement, but just as an alternative implementation: Would it not be better here to have this be `std::vector<std::pair<Addr, size_t>> &ranges`?

It saves any issues of people putting the beginning of the range before the end in the pairs, and makes figuring out the size of the of the logical  memory simple (sum all the sizes). I think you could also create a very simple algorithm for translating physical to logical addresses this way too. 

You'd only need to worry about the case where two ranges overlap (in which case  they should be one singular range), though this is still an issue with the current pair<start, end> design.

powerjg replied:
I think your proposal is fine, but I believe this design was chosen because it matches how AddrRanges were (and are) defined. They have always been defined with a begin and an end. The begin + size is a specialization. I'd prefer to keep the logic as it is.

BobbyRBruce replied:
Assumed this was the case. Thank you for the response. I'm ok with beginning and end since it's the de-facto historic way of doing it

BobbyRBruce commented on src/python/m5/params/param_types.py:
```diff
@@ -658,6 +658,124 @@ def is_subset(self, addr_range):
+
+
+class SparseModuloAddrRange(ModuloAddrRange):
+    def __init__(self, ranges, **kwargs):
```
Typing here would be nice. It's not clear what data type should be used for "ranges"

BobbyRBruce commented on src/base/addr_range.hh:
```diff
@@ -281,7 +404,86 @@ class AddrRange
+     *
+     * @ingroup api_addr_range
+     */
+    std::vector<std::pair<Addr, Addr>>
+    subRanges() const
+    {
+        if (isSparse()) {
+            return _chunks;
+        }
+        return std::vector<std::pair<Addr, Addr>>{{_start, _end}};
+    }
```
So, why is the addr_range being sparse or not so important; you to design everything so differently dependent on that. This code here just models a non-sparse address range as a sparse one with just one chunk. Can't we just have "chunks" be the central data structure here with just non-sparse case just having `Len(chunks) == 1`. It seems simpler to me

powerjg replied:
Yeah, except that performance is quite important here. I don't feel comfortable making the common path through the sparse logic.

BobbyRBruce replied:
Fair enough.

BobbyRBruce commented on src/base/addr_range.hh:
```diff
@@ -281,7 +404,86 @@ class AddrRange
+    bool
+    isSparse() const
+    {
+        return !_chunks.empty();
```
what would it mean here if `len(_chunks) == 1`?

BobbyRBruce replied:
I'll expand on this to say: buy that I mean, can we have a just one chunk?

BobbyRBruce reviewed (commented):
No major complaints here. I have some comments/questions/suggestions but everything seems, on the whole, ok.

abmerop reviewed (commented):
No complaints from me.  All our GPU tests pass.

powerjg committed:
base,mem: Refactor AddrRange for non-power-of-two and sparse

This commit refactors the AddrRange class to support non-power-of-two
sizes and sparse address ranges. This is necessary for supporting
modern memory systems that don't adhere to strict power-of-two sizes.

Key changes:
- Introduced AddrRangePolicy to encapsulate different mapping and
  interleaving strategies (e.g., Modulo, Sparse).
- Updated AddrRange to use AddrRangePolicy instead of hardcoding
  interleaving logic.
- Added support for sparse ranges where valid addresses are not
  contiguous.
- Updated PhysicalMemory and AbstractMemory to correctly handle
  lookups and checks against these new range types.
- Added extensive unit tests in src/base/addr_range.test.cc and
  src/base/addr_range_map.test.cc to verify correctness of
  interleaving, exclusion, and sparse mapping logic.
- Exposed new parameter types to Python in param_types.py and
  updated pybind11 bindings in core.cc to make these features
  accessible from Python scripts.

powerjg committed:
stdlib: Update components for refactored memory

This commit updates the gem5 standard library components to be
compatible with the refactored AddrRange and its new policies.

Key changes:
- Updated x86_board.py to handle new memory mapping requirements.
- Modified hbm.py to use sparse ranges or non-power-of-two sizes if
  needed, fixing instantiation issues with HBM stacks.
- Updated memory.py to correctly instantiate memory controllers
  with the updated address range definitions.
- Updated x86_demo_board.py to ensure it still works with the
  updated memory system.

powerjg committed:
configs: Update example scripts for refactored memory

This commit updates example scripts in configs/example/gem5_library/
to be compatible with the refactored AddrRange and stdlib changes.

Key changes:
- Updated x86-gapbs-benchmarks.py, x86-npb-benchmarks.py,
  x86-spec-cpu2006-benchmarks.py, and x86-spec-cpu2017-benchmarks.py
  to work with the updated memory configurations.
- Updated x86-ubuntu-run-with-kvm-no-perf.py and
  x86-ubuntu-run-with-kvm.py to ensure they instantiate correctly
  with the new address range logic.
- Fixed arguments and range specifications where necessary to match
  the new AddrRange requirements.

powerjg commented:
I have rebased these commits and squashed/reordered to clean things up. It is READY FOR REVIEW!

The old branch can be found here: https://github.com/powerjg/gem5/tree/old-refactor-addr-range

powerjg commented:
@giactra I would like to get this merged this week, if possible.

powerjg commented:
clang-format is passing locally. I am not sure what the problem is.

abmerop commented on src/base/addr_range.hh:
```diff
@@ -814,51 +1190,61 @@ operator-(const AddrRangeList &base, const AddrRangeList &to_exclude)
-}
+static inline AddrRange
+RangeSize(Addr start, Addr size)
+{ return AddrRange(start, start + size); }
```
example

abmerop reviewed (commented):
Maybe run clang-format directly? The clang-format hook runs for me if I soft reset to HEAD~1 and readd everything. It is mostly changing the one line method definitions

giactra reviewed (commented):
Hi @powerjg, thank you so much for this long-awaited change. I am really looking forward for this to be merged and I plan to dedicate more time in the reviewing process to fast track this.

I didn't look into the details of the patch, just into the overall design. These are my impressions:

1) It looks in good shape (other than some clang-format quirks)
2) I approve the interleaving policy and IMO once it gets a quick functionality check it is ready to be merged
3) About sparse ranges, this is also strongly needed in gem5. However as I believe I mentioned some time ago, I don't believe having sparse ranges implemented within the single AddrRange is the way to go.

To me it is much more maintainable if we "simply" tweaked AddrRange components to support vectors of AddrRange instead of encoding vectors of AddrRange within the AddrRange iteself.

In this way we keep the crucial AddrRange class simple, and rely more on composition of easy to reason building blocks. I believe this will make things much more maintenable. What do you think? Am I missing something?

powerjg commented:
Hi @giactra,

Thanks for the high-level architectural feedback! I completely understand the appeal of keeping `AddrRange` as a simple, singular building block and using composition (`std::vector<AddrRange>`) at the component level.

During the early stages of this design, I actually prototyped a composition approach, but ran into a few hard engineering constraints that forced the current embedded structure:

1. **Interleaving / Sparsity Interaction:** The implementation introduces complex sparse-interleaved types (like `SparseModuloAddrRange`). The interleaving policies need to evaluate a contiguous logical address space across all valid chunks while gracefully ignoring the holes. If we decouple these into a vector of independent `AddrRange` objects, the interleaving math becomes incredibly fragmented and forces individual components (like crossbars or controllers) to manually stitch the chunks together to compute channel matches.
2. **Interface Stability:** Moving to a vector approach would require modifying core simulator interfaces—such as changing packet routing targets and port range queries to handle collections of ranges rather than a unified pass-by-value token.

Interestingly, the current implementation actually does use composition where it matters most. For instance, in `AddrRangeMap::insert` and `PhysicalMemory::createBackingStore`, we use `r.decompose()` and `r.subRanges()` to unpack the sparse chunks and treat them as independent contiguous blocks at the structural layer. The single `AddrRange` class essentially acts as a clean interface wrapper around this behavior.

Given that this design satisfies both the encapsulation of the interleaving math and has successfully undergone rigorous functional and performance testing, changing the top-level architecture now would be incredibly costly.

What are your thoughts on merging this implementation as-is to unblock the community? If we identify areas down the road where decoupling the container makes sense, we can absolutely tackle that as an isolated, post-merge refactor.

giactra commented on src/base/addr_range.test.cc:
```diff
@@ -1594,8 +1595,335 @@ TEST(AddrRangeDeathTest, ExcludeInterleavingRanges)
+    EXPECT_TRUE(r0.contains(0));
+    EXPECT_TRUE(r0.contains(3));
+    EXPECT_FALSE(r0.contains(1));
+    EXPECT_FALSE(r0.contains(2));
+
+    EXPECT_TRUE(r1.contains(1));
+    EXPECT_TRUE(r1.contains(4));
```
Nit: for simmetry you should be checking all values from 0 to 5

powerjg replied:
Great suggestion. I have updated TEST(AddrRangeTest, ModuloSimple) so that r0, r1, and r2 all verify values 0 through 5 for full symmetry.

giactra commented on src/base/addr_range_policy.hh:
```diff
@@ -0,0 +1,544 @@
+    virtual std::string to_string() const = 0;
+};
+
+class MaskedInterleavingPolicy : public AddrMapPolicy
```
Could you add some documentation / example here on how the match works?

powerjg replied:
Oops. This was accidentally dropped. Agreed. I have restored and adapted the detailed Doxygen documentation from upstream gem5's legacy AddrRange constructor, which provides a concrete multi-bit mask example explaining how the XOR parity calculations form sel to compare against intlvMatch.

giactra commented on src/base/addr_range_policy.hh:
```diff
@@ -0,0 +1,544 @@
+    /**
+     * Compare policies for sorting.
+     */
+    virtual bool
+    lessThan(const std::shared_ptr<AddrMapPolicy> &other) const
+    {
+        auto ptr = other.get();
+        if (typeid(*this) != typeid(*ptr)) {
+            return typeid(*this).before(typeid(*ptr));
+        }
+        return false;
+    }
```
This seems very obscure. Could you guide me through its intended behaviour? Is it simply a way to tie-break ordering between ranges with different interleaving policies?

powerjg replied:
You hit the nail on the head. For STL containers like AddrRangeMap (and sets/maps using AddrRange as a key), strict weak ordering (operator<) requires a deterministic tie-breaker when comparing two ranges that have identical start and end addresses but use different policy classes. I have added a clarifying C++ comment to lessThan explaining this behavior.

giactra commented on src/mem/abstract_mem.cc:
```diff
@@ -112,7 +122,20 @@ AbstractMemory::setBackingStore(uint8_t* pmem_addr)
     // The back door can't handle interleaved memory.
     backdoor.ptr(range.interleaved() ? nullptr : pmem_addr);
```
Shouldn't we invalidate the backdoor for sparse ranges as well?

powerjg replied:
Yes, you are completely correct. PhysicalMemory::createBackingStore allocates separate host memory buffers for each subrange of a sparse range and calls setBackingStore multiple times. Previously, backdoor.ptr() got overwritten with the host buffer of the last subrange, which breaks MemBackdoor's contiguous linear offset calculation (paddr - range.start()). I have updated setBackingStore to invalidate the backdoor for both interleaved and sparse ranges: backdoor.ptr((range.interleaved() || range.isSparse()) ? nullptr : pmem_addr);.

giactra commented on src/mem/abstract_mem.hh:
```diff
@@ -117,6 +118,18 @@ class AbstractMemory : public ClockedObject
+    AddrRangeMap<uint8_t *> pmemMap;
+
+    // Whether the memory is sparse
+    const bool isSparse;
```
We could really collapse this into

pmemMap only.

I don't see the reason to keep pmemAddr (other than for some sort of compatibility), considering it increases the chances of using the wrong pointer. isSparse could be inferred from the fact that pmemMap.size() != 1. (I am fine on keeping it anyway if it is on the critical path)

powerjg replied:
You are completely right on the intuition, but pmemAddr is indeed on the absolute critical path of simulation memory accesses (functional, atomic, and packet accesses). Direct pointer calculation pmemAddr + addr - range.start() is O(1), whereas searching an AddrRangeMap (an interval tree) adds measurable latency during standard contiguous memory simulations. I have added a comment in the header documenting this critical path optimization.

[Jason, not AI] IIRC, pmemAddr is directly accessed in places other than `access`, but I could be mistaken. This is performance critical, so I was hesitant to to put the map access on the critical path.

giactra replied:
Ok, I understand the critical path optimization. However I have two comments:

1) I wonder if the slowdown has been actually measured (I am not skeptical, just curious)
2) I would shift the perspective and consider the contiguous range pmemAddr usage as a sort of optimization over the common case of sparse range. This effectively means that pmemMap is **ALWAYS** valid and initialized (even in the contiguous range case). This makes debugging easier as we rely on one data structure only to gather abstract memory range info.
We simply **cache** the single pointer into pmemAddr in order for it to be used in the critical path.

This seems more robust to me. I am really scared of having a potentially uninitialized pmemMap

powerjg replied:
> I wonder if the slowdown has been actually measured (I am not skeptical, just curious)

I did not directly test it. However, as an indirect test, https://github.com/gem5/gem5/issues/2855#issuecomment-3710977555 shows the performance with dual channel memory with sparse and not sparse address ranges. When using sparse ranges, the time spent in `AddrRangeMap` goes up by 3x. 

The `AddrRangeMap` code is not trivial, and putting it on the critical path would definitely cause a performance hit. If it's 10% or 5x, I have no idea.

> We simply cache the single pointer into pmemAddr in order for it to be used in the critical path.

This is a good approach. Let me investigate the cost of the implementation (likely low).

giactra commented on src/mem/physical.cc:
```diff
@@ -113,6 +113,17 @@ PhysicalMemory::PhysicalMemory(const std::string &_name,
+            } else {
+                validAddrMap.emplace_back(m->getAddrRange().start(),
+                                          m->getAddrRange().end());
+            }
```
Shouldn't subRanges work for a non sparse address range as well?

powerjg replied:
Excellent observation! AddrRange::subRanges() already returns {{_start, _end}} for non-sparse ranges, so the isSparse() check here was entirely redundant. I have removed the if/else check entirely and simplified it to unconditionally use subRanges().

giactra commented:
> Given that this design satisfies both the encapsulation of the interleaving math and has successfully undergone rigorous functional and performance testing, changing the top-level architecture now would be incredibly costly.
> 
> What are your thoughts on merging this implementation as-is to unblock the community? If we identify areas down the road where decoupling the container makes sense, we can absolutely tackle that as an isolated, post-merge refactor.

Sounds good to me. To me a possible (I haven't given too much thought to it) future development could be to define a smaller AddrRange type which is the contiguous version of the new AddrRange. To be used for simple components where we don't expect interleaving

powerjg reviewed (commented):
> Sounds good to me. To me a possible (I haven't given too much thought to it) future development could be to define a smaller AddrRange type which is the contiguous version of the new AddrRange. To be used for simple components where we don't expect interleaving

I thought about this, but I didn't come up with a simple way to do it. Somewhere we need polymorphism. Either we have it in the AddrRange (what I did), we make AddrRange an abstract class (this would require cascading changes all over the codebase), or we put the polymorphism at every AddrRange call site. I felt like all 3 were imperfect choices.

Now that we have a existence proof of new kinds of address ranges, refactoring this (e.g., with AI), should be straightforward. However, before we do that we need to make sure we have good test coverage, of course.

Thanks for you review. You caught a bunch of good stuff! Note that most of the responses are AI. I carefully reviewed the output though, so blame me if anything doesn't meet your expectations.

powerjg committed:
mem: Always populate pmemMap even for contiguous ranges

This commit addresses feedback by ensuring that pmemMap is always valid
and populated, even when the memory range is contiguous and non-sparse.
This avoids potential uninitialized pmemMap issues and simplifies
debugging. pmemAddr acts as a cached optimization for the contiguous case on
the simulation critical path.

powerjg commented:
~~On the clang format thing... here is what I did... I am at a total loss on what it's issue is. Every single commit said that the presubmit clang format passed.~~

I FINALLY FIGURED IT OUT! Locally, I have a different version of clang-format. The solution is to use `pip install clang-format==18.1.8` and then run the following.

`git rebase -i develop` and set all commits to `edit`. Then,

1. **Undo the current commit** (keeping all changes in the working tree):
   ```bash
   git reset HEAD~1
   ```

2. **Stage only the files that were part of the original commit**:
   ```bash
   git diff --name-only HEAD ORIG_HEAD | xargs git add
   ```
   *(This dynamically gets the list of files changed between the parent commit `HEAD` and the original commit `ORIG_HEAD`, and stages only those files).*

3. **Re-commit the changes** (triggering the hooks):
   ```bash
   git commit -c ORIG_HEAD
   ```

4. **Handle pre-commit hook modifications** (if any):
   If the clang-format hook automatically formats files, stage those formatting updates and commit:
   ```bash
   git add -u
   git commit -c ORIG_HEAD
   ```

5. **Continue the rebase**:
   ```bash
   git rebase --continue
   ```
   
Furthermore, the error says:
 
```
Command 'git-clang-format --style=file --commit=c2c505f97bfae67f71769505adea56bfe104f0c7' returned 1
```
 
This is what I see locally. It completes with exit code 0.
```
 % git-clang-format --style=file --commit=c2c505f97bfae67f71769505adea56bfe104f0c7
clang-format did not modify any files
```

powerjg commented:
@giactra thanks again for the diligent reviews. I believe I have addressed all of the concerns (and clang-format is finally passing!!!). I tried to mark things resolved where appropriate, and my intention is that everything is addressed. Let me know otherwise.

powerjg commented on src/base/addr_range.hh:
```diff
@@ -325,7 +549,21 @@ class AddrRange
     Addr
     size() const
     {
-        return (_end - _start) >> masks.size();
+        Addr total = 0;
+        if (isSparse()) {
+            for (const auto &chunk : _chunks) {
+                total += (chunk.second - chunk.first);
+            }
+        } else {
+            total = _end - _start;
+        }
+
+        if (_policy) {
+            // Policy assumes a contiguous logic space of 'total' size
+            // returning 'size' usually just divides by stripes
+            return _policy->size(0, total);
+        }
+        return total;
```
Done.

powerjg commented on src/python/m5/params/param_types.py:
```diff
@@ -658,6 +659,129 @@ def is_subset(self, addr_range):
+            self.intlvLowBit,
+        )
+
+    def size(self):
+        def to_compact(a):
+            return (
+                (a >> self.intlvLowBit) // self.stripes
+            ) << self.intlvLowBit | (a & ((1 << self.intlvLowBit) - 1))
+
+        return to_compact(int(self.end)) - to_compact(int(self.start))
+
```
This and the following comments are addressed by disallowing ranges that are not evenly divisible.

powerjg committed:
base: Fix hash helpers test compilation under GCC 15

giactra commented on src/base/addr_range.hh:
```diff
@@ -182,13 +181,168 @@ class AddrRange
+     *
+     * @ingroup api_addr_range
+     */
+    AddrRange(Addr start, Addr end, uint32_t stripes, uint32_t intlv_match,
+              uint32_t intlv_low_bit = 0)
```
Proposal: to avoid confusion between XOR based and Module based interleaving, 

example (numbers are random:

AddrRange(0, 30, 1, 2, 3) vs -> MODULO

AddrRange(0, 30, 1, 2, 3, 4) -> XOR

would relying on a factory method make things more readable? I mean making the module constructor protected/private and have somthing like

range = AddrRange::with_modulo_intlv(...)

We could have a symmetric version for the xor interleaving, but I would keep the constructor public for backwards compatibility

With the pybind interface (python) we could have something similar or we could rely on the py::kw_only option to make sure modulo data can only be passed as a KW argument (named) therefore avoding any ambiguity

https://pybind11.readthedocs.io/en/stable/advanced/functions.html

powerjg replied:
This is a great proposal. I am not going to be able to get to it for a bit (at least a week). We can wait to merge this until I get around to it, or we can merge it and I will create a new PR with this implementation. It's up to you. 🙂

giactra reviewed (commented):
Hi @powerjg, thanks for addressing my comments. It looks in a good state now. I am happy to approve this great change; I only have a last minute proposal which I believe will make code much more readable

giactra reviewed (approved):
I am going to approve this considering the importance of the PR. However I feel we should try to move quickly to the factory based API before users start adopting this temporary interface
